khtools/extract_coding.py

```diff
@@ -0,0 +1,510 @@
+        click.echo(f"\tK-mers in peptide database:", err=True)
+        click.echo(kmers_in_peptide_db, err=True)
+
+    fraction_in_peptide_db = n_kmers_in_peptide_db / n_kmers
+
+    return fraction_in_peptide_db, n_kmers
+
+
+def evaluate_is_fastp_low_complexity(seq, complexity_threshold=0.3):
+    """Use fastp's definition of complexity
+
+    fastp prpject: https://github.com/OpenGene/fastp
+
+    Low complexity = Many runs of the same base in a row
+    """
+    complexity = compute_fastp_complexity(seq)
+    return complexity < complexity_threshold
+
+
+def compute_fastp_complexity(seq):
+    n_different_consecutively = sum(1 for i in range(len(seq) - 1)
+                                    if seq[i] != seq[i + 1])
+    complexity = n_different_consecutively / len(seq)
+    return complexity
+
+
+def evaluate_is_kmer_low_complexity(sequence, ksize):
+    """Check if sequence is low complexity, i.e. mostly repetitive
+
+    By this definition, the sequence is not complex if its number of unique
+    k-mers is smaller than half the number of expected k-mers
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings('ignore')
+        # Ignore Biopython warning of seq objects being strings now
+        try:
+            kmers = kmerize(sequence, ksize)
+        except ValueError:
+            # k-mer size is larger than sequence
+            return None, None
+    n_kmers = len(kmers)
+    n_possible_kmers_on_sequence = len(sequence) - ksize + 1
+    min_kmer_entropy = n_possible_kmers_on_sequence / 2
+    if n_kmers <= min_kmer_entropy:
+        return True, n_kmers
+    return False, n_kmers
+
+
+def score_single_read(sequence,
+                      peptide_bloom_filter,
+                      peptide_ksize,
+                      molecule='protein',
+                      verbose=True,
+                      jaccard_threshold=0.9,
+                      description=None,
+                      noncoding_file_handle=None,
+                      coding_nucleotide_file_handle=None,
+                      low_complexity_peptide_file_handle=None):
+    # Convert to BioPython sequence object for translation
+    seq = Seq(sequence)
+
+    # Convert to BioPython sequence object for translation
+    translations = six_frame_translation_no_stops(seq)
+    # For all translations, use the one with the maximum number of k-mers
+    # in the databse
+    max_n_kmers = 0
+    max_fraction_in_peptide_db = 0
+    if len(translations) == 0:
+        return np.nan, np.nan, "No translation frames without stop codons"
+
+    translations = {
+        frame: translation
+        for frame, translation in translations.items()
+        if len(translation) > peptide_ksize
+    }
+    if len(translations) == 0:
+        return np.nan, np.nan, "All translations shorter than peptide k-mer " \
+                               "size + 1"
+
+    for frame, translation in translations.items():
+        if len(translation) <= peptide_ksize:
```

Comment:
this would never happen, right as you just filtered the translations dictionary above between lines 180-184?